File: 200.Number_of_Islands.cc
#include <iostream>
#include <cassert>
#include <vector>
#include <unordered_set>

class Solution {
private:
    std::vector<std::vector<int>> directions = {
        {1,0},
        {-1,0},
        {0,1},
        {0,-1}
    };
    int n = 0;
    int m = 0;
public:
    void dfs(int row, int col, std::vector<std::vector<bool>>& visited, std::vector<std::vector<char>>& grid) {
        if (row < 0 || row >= m || col < 0 || col >= n || grid[row][col] == '0' || visited[row][col]) {
            return;
        }

        visited[row][col] = true;;
        for (auto& direction: directions) {
            dfs(row + direction[0], col + direction[1], visited,grid);
        }
    }

    int numIslands(std::vector<std::vector<char>>& grid) {
        m = grid.size();
        n = grid[0].size();
        std::vector<std::vector<bool>> visited(m , std::vector<bool>(n, false));
        int count = 0;

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    dfs(i, j, visited, grid);
                    ++count;
                }
            }
        }

        std::cout << count << "\n";
        return count;
    }

};

/*
    Flood-Fill algorithm:
    Use DFS on all cells. If a cell is 0, can return immediately
    Whenever a 1 is spotted, add the cell (row,col) to a visited set and call dfs on it.
        Note: Unordered_set of pairs doesn't work without a custom hash function
    Explore all adjacent cells to see if they are also 1, add those cells to visited

    Time Complexity -- O(m*n)
*/

int main() {
    Solution solution;
    std::vector<std::vector<char>> grid1 = {
        {'1','1','1','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','0'},
        {'0','0','0','0','0'}
    };
    int ans1 = 1;
    assert(solution.numIslands(grid1) == ans1);

    std::vector<std::vector<char>> grid2 = {
        {'1','1','0','0','0'},
        {'1','1','0','0','0'},
        {'0','0','1','0','0'},
        {'0','0','0','1','1'}
    };
    int ans2 = 3;
    assert(solution.numIslands(grid2) == ans2);

    std::cout << "\033[32mAll tests passed!\033[0m\n";
    return 0;
}

File: Python solutions/200. Number of Islands.py
from typing import List

class Solution:
    def __init__(self):
        self.directions = [
            [1,0],
            [-1,0],
            [0, 1],
            [0,-1]
        ]

    def numIslands(self, grid: List[List[str]]) -> int:
        visited = set()
        m = len(grid)
        n = len(grid[0])
        count = 0

        def dfs(row: int, col: int) -> None:
            if row < 0 or row >= m or col < 0 or col >= n or (row, col) in visited or grid[row][col] == "0":
                return

            visited.add((row, col))

            for direction in self.directions:
                dfs(row + direction[0], col + direction[1])

        for row_idx, row in enumerate(grid):
            for col_idx, cell in enumerate(row):
                if cell == "1" and (row_idx, col_idx) not in visited:
                    dfs(row_idx, col_idx)
                    count += 1

        return count

assert Solution(). == , "Test failed!"

print("\033[32mAll tests passed!\033[0m\n");